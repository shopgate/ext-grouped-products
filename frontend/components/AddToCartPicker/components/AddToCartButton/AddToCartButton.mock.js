import React, { Component } from 'react';

export default class AddToCartButton extends Component {
  static get propTypes() {
    return {}
  }

  constructor(props) {
    super(props);

    this.state = {
      showCheckmark: null,
    };
  }

  render() {
    return (
      <button
        data-test-id="addToCartButton"
        onClick={this.handleClick}
      >
    </button>
    );
  }
};
