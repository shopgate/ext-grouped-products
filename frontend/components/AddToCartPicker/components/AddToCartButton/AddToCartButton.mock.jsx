import React, { Component } from 'react';

/**
 * Mocked AddToCartButton Component.
 */
export default class AddToCartButton extends Component {
  /**
   * PropTypes getter.
   */
  static get propTypes() {
    return {};
  }

  /**
   * Component constructor.
   * @param {Object} props The component props.
   */
  constructor(props) {
    super(props);

    this.state = {
      // eslint-disable-next-line react/no-unused-state
      showCheckmark: null,
    };
  }

  /**
   * The render method.
   * @return {JSX}
   */
  render() {
    return (
      <button
        data-test-id="addToCartButton"
        onClick={this.handleClick}
      />
    );
  }
}
