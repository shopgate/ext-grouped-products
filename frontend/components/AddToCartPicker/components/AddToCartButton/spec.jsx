import React from 'react';
import { mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import PickerAddToCartButton from './index';

describe('<PickerAddToCartButton />', () => {
  const mockStore = configureStore();
  it('Should render in loadingState and should not be clickable', () => {
    const spy = jest.fn(() => new Promise(resolve => resolve()));
    const store = mockStore({});
    const wrapper = mount(<PickerAddToCartButton
      store={store}
      onClick={spy}
      openList={spy}
      isLoading
      isOrderable
      isDisabled={false}
      conditioner={{
        check: () => new Promise(resolve => resolve(true))
      }}
    />);
    expect(wrapper).toMatchSnapshot();
    expect(spy).toHaveBeenCalledTimes(0);
  });

  it('should render with checkmark icon and should be clickable', () => {
    const spy = jest.fn(() => new Promise(resolve => resolve()));
    const store = mockStore({});
    const wrapper = mount(<PickerAddToCartButton
      store={store}
      onClick={spy}
      openList={spy}
      isLoading={false}
      isOrderable
      isDisabled={false}
      conditioner={{
        check: () => new Promise(resolve => resolve(true))
      }}
    />);
    wrapper.find('button').prop('onClick')();
    wrapper.update();

    expect(wrapper).toMatchSnapshot();
    expect(spy).toHaveBeenCalledTimes(1);
  });
});
