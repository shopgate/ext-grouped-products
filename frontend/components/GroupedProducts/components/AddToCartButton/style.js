import { css } from 'glamor';

const button = css({
  boxShadow: 'none !important',
}).toString();

export default {
  button,
};
