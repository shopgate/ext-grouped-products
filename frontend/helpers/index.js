import { generateResultHash } from '@shopgate/pwa-common/helpers/redux';
import appConfig from '@shopgate/pwa-common/helpers/config';
import { SHOPGATE_CATALOG_GET_PRODUCT_CHILDREN, THEME_GMD } from '../constants';
import { maxQuantityPickerEntries as maxEntries } from '../config';

/**
 * Generates a hash to read and write with the productsByHash store.
 * @param {string} productId A product id.
 * @return {string}
 */
export const generateHash = productId => generateResultHash({
  pipeline: SHOPGATE_CATALOG_GET_PRODUCT_CHILDREN,
  productId,
});

/**
 * Creates an item list for the AddToCartPicker.
 * @param {Object} stock The stock property of a product.
 * @return {Array}
 */
export const createPickerItems = (stock) => {
  if (!stock) {
    return [];
  }

  const {
    ignoreQuantity,
    quantity,
  } = stock;

  const items = [];

  for (let i = 0; i <= maxEntries; i += 1) {
    if (ignoreQuantity) {
      if (items.length === maxEntries) {
        // When the quantity can be ignored the maximum number of entries can be created.
        break;
      }
    } else if (items.length === maxEntries || i >= quantity) {
      // When the quantity can't be ignored the maxium number of entries is limited.
      break;
    }

    const value = i + 1;

    items.push({
      label: `${value}`,
      value,
    });
  }

  return items;
};

/**
 * Checks if the current theme is the GMD theme.
 * @return {boolean}
 */
export const isGmdTheme = () => process.env.THEME.includes(THEME_GMD);

/**
 * Tells if the extension shall render flat buttons.
 * @return {boolean}
 */
export const renderFlatButtons = () => !isGmdTheme();

/**
 * Check if the favorite list feature is activated.
 * @return {boolean}
 */
export const hasFavorites = () => appConfig.hasFavorites;
