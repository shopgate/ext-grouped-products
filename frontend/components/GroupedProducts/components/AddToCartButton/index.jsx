import React, { Component } from 'react';
import PropTypes from 'prop-types';
import AddToCartButton from '@shopgate/pwa-ui-shared/AddToCartButton';
import styles from './style';

/**
 * The AddToCartButton for grouped products.
 */
class GPAddToCartButton extends Component {
  static propTypes = {
    stock: PropTypes.shape().isRequired,
  };

  handleAddToCart = () => {

  };

  /**
   * The render method.
   * @return {JSX}
   */
  render() {
    const { orderable } = this.props.stock;
    const isDisabled = false;
    const isLoading = false;

    return (
      <AddToCartButton
        className={styles.button}
        isOrderable={orderable}
        isDisabled={isDisabled}
        isLoading={isLoading}
        handleAddToCart={this.handleAddToCart}
      />
    );
  }
}

export default GPAddToCartButton;
