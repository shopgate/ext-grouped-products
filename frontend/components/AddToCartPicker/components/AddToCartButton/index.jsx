import Event from '@shopgate/pwa-core/classes/Event';
import { logger } from '@shopgate/pwa-core/helpers';
import AddToCartButton from '@shopgate/pwa-ui-shared/AddToCartButton';
import { EVENT_ADD_TO_CART_MISSING_VARIANT } from '@shopgate/pwa-common-commerce/cart/constants';
import { ADD_TO_CART_BUTTON_TYPE_DEFAULT } from '../../../../constants';

/**
 * The PickerAddToCartButton component.
 */
class PickerAddToCartButton extends AddToCartButton {
  /**
   * Click handler for the button.
   * @return {boolean}
   */
  handleClick = () => {
    const { showCheckmark } = this.state;
    const {
      isDisabled,
      isLoading,
      isOrderable,
      openList,
      type,
    } = this.props;

    if (showCheckmark || isLoading || isDisabled) {
      return false;
    }

    if (!isOrderable) {
      if (type === ADD_TO_CART_BUTTON_TYPE_DEFAULT) {
        // Fire an event, so that the UI can react on the failed "add to cart" attempt.
        Event.call(EVENT_ADD_TO_CART_MISSING_VARIANT);
      }

      return false;
    }

    if (typeof openList === 'function') {
      this.props.openList();
    } else {
      logger.error('No openList() prop provided');
    }

    return true;
  }

  /**
   * Called when the added quantity was increased. It triggers the button animation.
   */
  onQuantityIncreased() {
    this.setState({
      showCheckmark: true,
    });

    setTimeout(() => {
      this.setState({
        showCheckmark: false,
      });
    }, 900);
  }
  /**
   * Component did update lifecycle hook.
   * @param {Object} prevProps The previous props.
   */
  componentDidUpdate(prevProps) {
    if (this.props.addedQuantity > prevProps.addedQuantity) {
      // Trigger the button animation when the quantity was increased.
      setTimeout(() => {
        this.onQuantityIncreased();
      }, 250);
    }
  }

  /**
   * Renders the extended component.
   * @return {JSX}
   */
  render() {
    return super.render();
  }
}

export default PickerAddToCartButton;
