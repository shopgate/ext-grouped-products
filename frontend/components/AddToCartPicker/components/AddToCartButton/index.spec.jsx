import React from 'react';
import { mount } from 'enzyme';
import { logger } from '@shopgate/pwa-core/helpers';
import AddToCartButton from './index';

jest.mock('@shopgate/pwa-core/helpers', () => ({
  logger: {
    error: jest.fn(),
  },
}));

const openListSpy = jest.fn();

const buttonProps = {
  handleAddToCart: () => {},
  openList: openListSpy,
  isDisabled: false,
  isLoading: false,
  isOrderable: true,
  addedQuantity: 0,
  noShadow: true,
};

jest.useFakeTimers();

describe('<AddToCartButton />', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render as expected', () => {
    const component = mount(<AddToCartButton {...buttonProps} />);
    expect(component).toMatchSnapshot();
  });

  it('should execute the click handler as expected', () => {
    const component = mount(<AddToCartButton {...buttonProps} />);
    /* eslint-disable extra-rules/no-single-line-objects */
    component.simulate('click');
    expect(openListSpy).toHaveBeenCalledTimes(1);

    component.setProps({ ...buttonProps, isDisabled: true });
    component.simulate('click');
    expect(openListSpy).toHaveBeenCalledTimes(1);

    component.setProps({ ...buttonProps, isOrderable: false });
    component.simulate('click');
    expect(openListSpy).toHaveBeenCalledTimes(1);

    component.setProps({ ...buttonProps, isLoading: true });
    component.simulate('click');
    expect(openListSpy).toHaveBeenCalledTimes(1);

    // Trigger the showCheckmark state change.
    component.setProps({ ...buttonProps, addedQuantity: 4 });
    component.update();
    jest.runOnlyPendingTimers();
    component.simulate('click');
    expect(openListSpy).toHaveBeenCalledTimes(1);
    jest.runOnlyPendingTimers();

    // Enable the click again for verification.
    component.setProps({ ...buttonProps });
    component.simulate('click');
    expect(openListSpy).toHaveBeenCalledTimes(2);
    /* eslint-enable extra-rules/no-single-line-objects */
  });

  it('should log an error if no openList handler was injected', () => {
    const component = mount(<AddToCartButton {...buttonProps} openList={null} />);
    component.simulate('click');

    expect(openListSpy).not.toHaveBeenCalled();
    expect(logger.error).toHaveBeenCalledTimes(1);
  });
});
