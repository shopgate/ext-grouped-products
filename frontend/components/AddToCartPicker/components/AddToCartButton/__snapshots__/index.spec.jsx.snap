// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddToCartButton /> should render as expected 1`] = `
<PickerAddToCartButton
  addedQuantity={0}
  handleAddToCart={[Function]}
  isDisabled={false}
  isLoading={false}
  isOrderable={true}
  noShadow={true}
  openList={[MockFunction]}
>
  <button
    data-test-id="addToCartButton"
    onClick={[Function]}
  />
</PickerAddToCartButton>
`;
